### PR TITLE
Add EC Mul and Add acceleration

### PR DESCRIPTION
**Pull request by tzerrell**
Adds elliptic curve acceleration via bigint2 for addition and scalar multiplication. This also includes a fix to allow bigint2 codegen with negative intermediates.

Note that current support is only for secp256k1; more generic curve support is future work.

### CODE REVIEW

File: risc0/bigint2/methods/guest/src/bin/ec_mul.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+#[allow(unused)]
+use risc0_zkvm::guest::env;
+
+use risc0_bigint2::ec::AffinePt;
+
+#[cfg(feature = "num-bigint")]
+use num_bigint::BigUint;
+#[cfg(feature = "num-bigint-dig")]
+use num_bigint_dig::BitUint;
```
Comment (flaub) [inline risc0/bigint2/methods/guest/src/bin/ec_mul.rs:23]: This should be un-needed, it's covered by this in `lib.rs`:

```
#[cfg(feature = "num-bigint-dig")]
extern crate num_bigint_dig as num_bigint;
```

File: risc0/bigint2/methods/guest/src/bin/ec_mul.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    let expected = AffinePt {
+        x: BigUint::from_slice(&[
+            0xfc345d74, 0xf1c13eb1, 0x0e1498e2, 0x881d811e,
+            0xd64702ef, 0xd73df930, 0x6ee88cbb, 0x77f23093
+        ]),
+        y: BigUint::from_slice(&[
+            0x671c60d6, 0xbe8eb3c7, 0xd97077cb, 0x96c95330,
+            0x9ba1b378, 0x0a08266e, 0x7886b640, 0x958ef42a
+
+        ])
```
Comment (flaub) [inline risc0/bigint2/methods/guest/src/bin/ec_mul.rs:46]: cargo fmt

File: risc0/bigint2/src/ec/mod.rs [hunk 3]
```diff
@@ -17,12 +17,126 @@ mod tests;
...
+use num_bigint_dig::BigUint;
 
+const ADD_BLOB: &[u8] = include_bytes_aligned!(4, "add.blob");
 const DOUBLE_BLOB: &[u8] = include_bytes_aligned!(4, "double.blob");
 
-pub fn double(point: &[u32], result: &mut [u32]) {
+pub const EC_256_WIDTH_WORDS: usize = 256 / 32;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AffinePt {
```
Comment (flaub) [inline risc0/bigint2/src/ec/mod.rs:32]:
```suggestion
pub struct AffinePoint {
```

File: risc0/bigint2/src/ec/mod.rs [hunk 4]
```diff
@@ -17,12 +17,126 @@ mod tests;
 
 use include_bytes_aligned::include_bytes_aligned;
 
-use crate::ffi::sys_bigint2_2;
+use crate::ffi::{sys_bigint2_2, sys_bigint2_3};
+#[cfg(not(feature = "num-bigint-dig"))]
+use num_bigint::BigUint;
+#[cfg(feature = "num-bigint-dig")]
+use num_bigint_dig::BigUint;
```
Comment (flaub) [inline risc0/bigint2/src/ec/mod.rs:24]: This should be un-needed, it's covered by this in `lib.rs`:

```
#[cfg(feature = "num-bigint-dig")]
extern crate num_bigint_dig as num_bigint;
```

File: risc0/bigint2/methods/guest/src/bin/ec_mul.rs (hunk 1 above)
Comment (flaub) [inline risc0/bigint2/methods/guest/src/bin/ec_mul.rs:23]:
```suggestion
#[cfg(feature = "num-bigint-dig")]
extern crate num_bigint_dig as num_bigint;
```

File: risc0/bigint2/methods/guest/src/bin/ec_add.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+#[allow(unused)]
+use risc0_zkvm::guest::env;
+
+use risc0_bigint2::ec::AffinePt;
+
+#[cfg(feature = "num-bigint")]
+use num_bigint::BigUint;
+#[cfg(feature = "num-bigint-dig")]
+use num_bigint_dig::BitUint;
```
Comment (flaub) [inline risc0/bigint2/methods/guest/src/bin/ec_add.rs:23]:
```suggestion
#[cfg(feature = "num-bigint-dig")]
extern crate num_bigint_dig as num_bigint;```

**Review by flaub — APPROVED** [review]

File: risc0/bigint2/src/ec/mod.rs [hunk 6]
```diff
@@ -16,13 +16,124 @@
...
+        let mut first = self.x.to_u32_digits();
+        assert!(first.len() <= EC_256_WIDTH_WORDS);
+        if first.len() < EC_256_WIDTH_WORDS {
+            first.resize(EC_256_WIDTH_WORDS, 0);
+        }
+        let mut last = self.y.to_u32_digits();
+        assert!(last.len() <= EC_256_WIDTH_WORDS);
+        if last.len() < EC_256_WIDTH_WORDS {
+            last.resize(EC_256_WIDTH_WORDS, 0);
+        }
```
Comment (flaub) [inline risc0/bigint2/src/ec/mod.rs:53]: We should try to avoid resizes

Comment (tzerrell) [reply risc0/bigint2/src/ec/mod.rs:53]: :+1: I think I can drop these in the generalized EC architecture, I'll make sure to pay attention to that

**Pull request merged**